server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -614,6 +658,88 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+        Map<String, DataStream> dataStreams;
+        List<String> requestedDataStreams = filterIndices(snapshotInfo.dataStreams(), requestIndices.toArray(String[]::new),
+            IndicesOptions.fromOptions(true, true, true, true));
+        if (requestedDataStreams.isEmpty()) {
+            dataStreams = new HashMap<>();
+        } else {
+            if (globalMetadata == null) {
+                globalMetadata = repository.getSnapshotGlobalMetadata(snapshotId);
+            }
+            final Map<String, DataStream> dataStreamsInSnapshot = globalMetadata.dataStreams();
+            dataStreams = new HashMap<>(requestedDataStreams.size());
+            for (String requestedDataStream : requestedDataStreams) {
+                final DataStream dataStreamInSnapshot = dataStreamsInSnapshot.get(requestedDataStream);
+                assert dataStreamInSnapshot != null : "DataStream [" + requestedDataStream + "] not found in snapshot";
+                dataStreams.put(requestedDataStream, dataStreamInSnapshot);
+            }
+        }
+        return dataStreams;
+    }
+
+    private Map<String, List<String>> getFeatureStatesToRestore(RestoreSnapshotRequest request, SnapshotInfo snapshotInfo,
+                                                                Snapshot snapshot) {
+        if (snapshotInfo.featureStates() == null) {
+            return Collections.emptyMap();
+        }
+        final Map<String, List<String>> snapshotFeatureStates = snapshotInfo.featureStates().stream()
+            .collect(Collectors.toMap(SnapshotFeatureInfo::getPluginName, SnapshotFeatureInfo::getIndices));
+
+        final Map<String, List<String>> featureStatesToRestore = new HashMap<>(snapshotFeatureStates);
+        if (request.featureStates() != null) {
+            final Set<String> requestedStates = Set.of(request.featureStates());
+            if (snapshotFeatureStates.keySet().containsAll(requestedStates) == false) {
+                Set<String> nonExistingRequestedStates = new HashSet<>(requestedStates);
+                nonExistingRequestedStates.removeAll(snapshotFeatureStates.keySet());
+                throw new SnapshotRestoreException(snapshot, "requested feature states [" + nonExistingRequestedStates +
+                    "] are not present in snapshot");
+            }
+            featureStatesToRestore.keySet().retainAll(requestedStates);
+        } else if (request.includeGlobalState() == false) {
+            featureStatesToRestore.clear();
+        }
+
+        final List<String> featuresNotOnThisNode = featureStatesToRestore.keySet().stream()
+            .filter(featureName -> systemIndices.getFeatures().containsKey(featureName))
+            .collect(Collectors.toList());
+        if (featuresNotOnThisNode.isEmpty() == false) {
+            logger.warn("while restoring feature states from snapshot [{}], some requested feature states are not present on this node: {}",
```

Comment:
shouldn't this be an error and not a log message?

Reply:
Looking at this again, I agree, it probably should be an error. Not certain why I didn't make it one originally. I'll change this, although I'm not sure if it's possible to write a test for this case given how our test infrastructure works.